// .ignored/@divriots/jampack/dist/compressors/css.js
import browserslist from 'browserslist';
import { browserslistToTargets, transform as lightcss, transformStyleAttribute as lightcssStyleAttribute, } from 'lightningcss';
let targets = browserslistToTargets(browserslist('defaults'));
export async function compressCSS(originalCode, type) {
    // Compress with lightningcss
    let lightCSSData = undefined;
    try {
        const options = {
            code: originalCode,
            minify: true,
            sourceMap: false,
            targets,
        };
        if (type === 'inline') {
            lightCSSData = lightcssStyleAttribute(options).code;
        }
        else {
            lightCSSData = lightcss({
                filename: 'style.css',
                ...options,
            }).code;
        }
    }
    catch (e) {
        // Error while processing with lightningcss
        // Take original code
        // TODO catch SyntaxError and report a Warning
    }
    let resultBuffer = undefined;
    if (lightCSSData && lightCSSData.length < originalCode.length) {
        resultBuffer = lightCSSData;
    }
    return resultBuffer || originalCode;
}
export function loadConfig(config) {
    targets = browserslistToTargets(browserslist(config.css.browserslist || config.general.browserslist));
}
