// unpic@3.22.0/script/src/transformers/supabase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = exports.generate = exports.parse = void 0;
const utils_js_1 = require("../utils.js");
const ALLOWED_FORMATS = ["origin"];
const STORAGE_URL_PREFIX = "/storage/v1/object/public/";
const RENDER_URL_PREFIX = "/storage/v1/render/image/public/";
const isRenderUrl = (url) => url.pathname.startsWith(RENDER_URL_PREFIX);
const parse = (imageUrl) => {
    const url = (0, utils_js_1.toUrl)(imageUrl);
    const isRender = isRenderUrl(url);
    if (!isRender) {
        return {
            cdn: "supabase",
            base: url.origin + url.pathname,
        };
    }
    const imagePath = url.pathname.replace(RENDER_URL_PREFIX, "");
    const quality = url.searchParams.has("quality")
        ? Number(url.searchParams.get("quality"))
        : undefined;
    const width = url.searchParams.has("width")
        ? Number(url.searchParams.get("width"))
        : undefined;
    const height = url.searchParams.has("height")
        ? Number(url.searchParams.get("height"))
        : undefined;
    const format = url.searchParams.has("format")
        ? url.searchParams.get("format")
        : undefined;
    const resize = url.searchParams.has("resize")
        ? url.searchParams.get("resize")
        : undefined;
    return {
        cdn: "supabase",
        base: url.origin + STORAGE_URL_PREFIX + imagePath,
        width,
        height,
        format,
        params: {
            quality,
            resize,
        },
    };
};
exports.parse = parse;
const generate = ({ base, width, height, format, params }) => {
    const parsed = (0, exports.parse)(base.toString());
    base = parsed.base;
    width = width || parsed.width;
    height = height || parsed.height;
    format = format || parsed.format;
    params = { ...parsed.params, ...params };
    const searchParams = new URLSearchParams();
    if (width)
        searchParams.set("width", (0, utils_js_1.roundIfNumeric)(width).toString());
    if (height)
        searchParams.set("height", (0, utils_js_1.roundIfNumeric)(height).toString());
    if (format && ALLOWED_FORMATS.includes(format)) {
        searchParams.set("format", format);
    }
    if (params?.quality) {
        searchParams.set("quality", (0, utils_js_1.roundIfNumeric)(params.quality).toString());
    }
    if (params?.resize)
        searchParams.set("resize", params.resize);
    if (searchParams.toString() === "")
        return base;
    return parsed.base.replace(STORAGE_URL_PREFIX, RENDER_URL_PREFIX) + "?" +
        searchParams.toString();
};
exports.generate = generate;
const transform = ({ url, width, height, format, cdnOptions }) => {
    const parsed = (0, exports.parse)(url);
    return (0, exports.generate)({
        base: parsed.base,
        width: width || parsed.width,
        height: height || parsed.height,
        format: format || parsed.format,
        params: cdnOptions?.supabase || parsed.params,
    });
};
exports.transform = transform;
